src/triage/component/catwalk/__init__.py

```diff
@@ -2,5 +2,141 @@
+                """,
+                    test_uuid,
+                )
+                continue
+            for train_task in train_tasks:
+                train_test_tasks.append(
+                    {
+                        "test_store": test_store,
+                        "train_store": train_store,
+                        "train_kwargs": train_task,
+                    }
+                )
+        return train_test_tasks
+
+    def process_all_tasks(self, tasks):
+        for task in tasks:
+            self.process_task(**task)
+
+    def process_task(self, test_store, train_store, train_kwargs):
+        logging.info("Beginning train task %s", train_kwargs)
+        model_id = self.model_trainer.process_train_task(**train_kwargs)
+        if not model_id:
+            logging.warning("No model id returned from ModelTrainer.process_train_task, "
+                            "training unsuccessful. Not attempting to test")
+            return
+        logging.info("Trained task %s and got model id %s", train_kwargs, model_id)
+        as_of_times = test_store.metadata["as_of_times"]
+        logging.info(
+            "Testing and scoring model id %s with test matrix %s. "
+            "as_of_times min: %s max: %s num: %s",
+            model_id,
+            test_store.uuid,
+            min(as_of_times),
+            max(as_of_times),
+            len(as_of_times),
+        )
+
+        self.individual_importance_calculator.calculate_and_save_all_methods_and_dates(
+            model_id, test_store
+        )
+
+        # Generate predictions for the testing data then training data
+        for store in (test_store, train_store):
+            if self.predictor.replace or self.model_evaluator.needs_evaluations(store, model_id):
+                logging.info(
+                    "The evaluations needed for matrix %s-%s and model %s"
+                    "are not all present in db, so predicting and evaluating",
+                    store.uuid,
+                    store.matrix_type,
+                    model_id
+                )
+                predictions_proba = self.predictor.predict(
+                    model_id,
+                    store,
+                    misc_db_parameters=dict(),
+                    train_matrix_columns=train_store.columns(),
+                )
+
+                self.model_evaluator.evaluate(
+                    predictions_proba=predictions_proba,
+                    matrix_store=store,
+                    model_id=model_id,
+                )
+            else:
+                logging.info(
+                    "The evaluations needed for matrix %s-%s and model %s are all present"
+                    "in db from a previous run (or none needed at all), so skipping!",
+                    store.uuid,
+                    store.matrix_type,
+                    model_id
+                )
+        self.model_trainer.uncache_model(model_id)
```

Comment:
It's not clear to me just by looking at this method where the model gets cached -- (I'm guessing it happens for each loop?) -- regardless, perhaps the method/block which caches and uses the model should be in a `try`, and this uncache expression in a `finally`? (For one, that might help protect against memory leaks.)

Reply:
It happens in model_trainer.process_train_task. We could certainly put a try/finally here

Reply:
Interesting! I'm going back and forth on whether or not the semantics of the context manager caching like that make sense....and I think I like it. Maybe in a similar vein there could be a `train_and_cache` that *has* to be run as a context manager, and use that instead.

I definitely agree that having the caching and the uncaching on different layers is bad. I'll play around and get them on the same layer with a context manager